Some adjustments for KEYCLOAK-3056 / PR #3893

skjolber opened the pull request:
In addtion to fixing/adding some unit tests, an additional test was added for documents actually having a signature when wantAssertionsSigned is enabled.

skjolber committed:
Add unit tests related to signatures, check that a signature is present when want assertion signing.

stianst commented:
Does this assume https://github.com/keycloak/keycloak/pull/3893 would be merged first? Or does it contain the changes?

Since you're both from Norway and are both interested in this is it safe to guess that you both work for the same company?

stianst commented:
Never mind. I somehow missed that 3893 was already merged.

skjolber commented:
I'll look into the unit test which are failing

anderius commented:
@stianst: No, but we might be integrating with the same IdP. Sorry I never got to complete the other PR, as the priority fell when we realised the IdP signed the response as well as the assertion. Then I went for fathers leave...

skjolber commented:
I am having some problems in recreating the failing unit tests locally, so I'll try a few (amended) commits and get some additional log output.

stianst commented:
As the unit tests was breaking our build I merged the fix here:
https://github.com/keycloak/keycloak/commit/f63c60855efd71427f678dc5ce73691937e3a6fb

Travis wasn't running unit tests and that's fixed now so this shouldn't happen again.

hmlnarik commented on services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java:
```diff
@@ -371,13 +371,12 @@ protected Response handleLoginResponse(String samlResponse, SAMLDocumentHolder h
-                        event.error(Errors.INVALID_SIGNATURE);
-                        return ErrorPage.error(session, Messages.INVALID_REQUESTER);
-                    }
+                if ((config.isWantAssertionsSigned() && !XMLSignatureUtil.isSigned(assertionElement.getOwnerDocument()))
```
WantAssertionsSigned requires signed assertions while what is checked for validity in isSigned is the whole document (arbitrary element signature would suffice to satisfy the condition in that method). This has to be fixed to reflect WantAssertionsSigned logic.

hmlnarik commented on saml-core/src/test/resources/org/keycloak/saml/processing/core/parsers/saml/v2/util/saml20-signed-response.xml:
```diff
@@ -0,0 +1 @@
+<samlp:Response xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol" xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion" Destination="http://localhost:8080/auth/realms/saml-broker-authentication-realm/broker/saml-identity-provider/endpoint" ID="ID_9aca1381-6265-434c-98a5-d89236d32ea0" InResponseTo="ID_2d81575a-0bde-488a-b69d-0b78c0fcf521" IssueInstant="2017-05-04T22:18:13.550Z" Version="2.0"><saml:I ...
```
This XML needs to be properly formatted, but actually see below.

hmlnarik commented on saml-core/src/main/java/org/keycloak/saml/processing/core/saml/v2/util/AssertionUtil.java:
```diff
@@ -272,24 +272,21 @@ public static void createSAML11TimedConditions(SAML11AssertionType assertion, lo
+    
     public static boolean isSignatureValid(Element assertionElement, KeyLocator keyLocator) {
         try {
             Document doc = DocumentUtil.createDocument();
```
Fix to the previous PR that introduced this method - this method should use XMLSignatureUtil.validateSingleNode() (make it public beforehand) instead of reimplementing the logic here.

hmlnarik commented on saml-core/src/main/java/org/keycloak/saml/processing/core/saml/v2/util/AssertionUtil.java:
```diff
@@ -557,21 +554,25 @@ public static AssertionType getAssertion(ResponseType responseType, PrivateKey p
     }
 
-    public static boolean isAssertionEncrypted(ResponseType responseType) throws ProcessingException {
+    public static boolean isAssertionsEncrypted(ResponseType responseType) throws ProcessingException {
```
Currently Keycloak only uses first assertion and this PR does not change this behaviour. Why to introduce this change?

hmlnarik commented on saml-core/src/main/java/org/keycloak/saml/processing/core/util/IDFedLSInputResolver.java:
```diff
@@ -113,7 +114,7 @@ public LSInput resolveResource(String type, String namespaceURI, final String pu
         if (systemId == null)
             throw logger.nullValueError("systemid");
-        if (StringUtil.isNotNull(systemId) && systemId.endsWith("dtd") && StringUtil.isNotNull(baseURI)) {
+        if (StringUtil.isNotNull(systemId) && !systemId.endsWith("dtd") && StringUtil.isNotNull(baseURI)) {
```
This change is unrelated to KEYCLOAK-3056 and lacks tests. Remove.

hmlnarik commented on saml-core/src/main/java/org/keycloak/saml/processing/core/util/IDFedLSInputResolver.java:
```diff
@@ -49,6 +49,7 @@
         // XML Schema/DTD
         schemaLocationMap.put("datatypes.dtd", "schema/w3c/xmlschema/datatypes.dtd");
         schemaLocationMap.put("XMLSchema.dtd", "schema/w3c/xmlschema/XMLSchema.dtd");
+        schemaLocationMap.put("http://www.w3.org/2001/XMLSchema.dtd", "schema/w3c/xmlschema/XMLSchema.dtd");        
```
This change is unrelated to KEYCLOAK-3056 and lacks tests. Remove.

skjolber replied:
The previous PR changed the SPMetadataDescriptor which generates XML. So the unit test ValidationTest was updated to actually schema-validate the generated output. However this did not work right without a few fixes to IDFedLSInputResolver. Constructing a dedicated test for the resolver is quite difficult, since it need to interfere with outgoing http connections. The best way is to just disconnect the internet and see that schema loading still works. Any ideas?

skjolber replied:
Added an unit test for IDFedLSInputResolver, was not so hard after all.

hmlnarik commented on services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java:
```diff
@@ -351,10 +351,10 @@ protected Response handleLoginResponse(String samlResponse, SAMLDocumentHolder h
                 }
 
-                boolean assertionIsEncrypted = AssertionUtil.isAssertionEncrypted(responseType);
+                boolean assertionIsEncrypted = AssertionUtil.isAssertionsEncrypted(responseType);
```
Currently Keycloak only uses first assertion and this PR does not change this behaviour. Why to introduce this change?

hmlnarik commented on services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java:
```diff
@@ -351,10 +351,10 @@ protected Response handleLoginResponse(String samlResponse, SAMLDocumentHolder h
 
                 if (config.isWantAssertionsEncrypted() && !assertionIsEncrypted) {
-                    logger.error("The assertion is not encrypted, which is required.");
+                    logger.error("The assertions are not encrypted, which is required.");
```
Currently Keycloak only uses first assertion and this PR does not change this behaviour. Why to introduce this change?

hmlnarik commented on services/src/main/java/org/keycloak/protocol/saml/SamlService.java:
```diff
@@ -580,7 +580,7 @@ public static String getIDPMetadataDescriptor(UriInfo uriInfo, KeycloakSession s
     }
 
-    private static void addKeyInfo(StringBuilder target, RsaKeyMetadata key, String purpose) {
+    protected static void addKeyInfo(StringBuilder target, RsaKeyMetadata key, String purpose) {
```
What is reason for this change?

skjolber replied:
Leftover from refactoring, removing.

hmlnarik reviewed (changes requested):
This PR currently touches several areas at once, including those that are not contributing to KEYCLOAK-3056. Please remove unrelated changes. Furthermore please remove tests with mocks.

skjolber commented:
I was unaware that Keycloak just uses the first assertion, so I'll remove those changes for multiple assertions. Properties name like wantAssertionsEncrypted got me confused.

hmlnarik commented:
That's true. Keycloak broker assertion processing only uses first assertion though, see https://github.com/skjolber/keycloak/blob/a074a17ce263b33acdcb2a200ca435f5081d4e26/services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java#L382

skjolber commented:
Removed mocking. Fixed unit test for IDFedLSInputResolver. Signature checks should now be for a single Assertion (as a direct child).

Was not able to pretty-print saml20-signed-response.xml without breaking signature check.

hmlnarik commented on testsuite/integration/src/test/java/org/keycloak/testsuite/saml/ValidationTest.java:
```diff
@@ -16,81 +16,85 @@
  */
```
Subject for another JIRA. Remove.

hmlnarik commented on saml-core/src/main/java/org/keycloak/saml/common/util/DocumentUtil.java:
```diff
@@ -554,4 +554,33 @@ private static DocumentBuilderFactory getDocumentBuilderFactory() {
+                String ns = childElement.getNamespaceURI();
+                String localName = childElement.getLocalName();
+                
+                if(targetNamespace.equals(ns) && targetLocalName.equals(localName)) {
```
Use ```Objects.equals(targetNamespace, ns)``` (dtto for local name) instead to prevent accidental NPE

hmlnarik commented on saml-core/src/main/java/org/keycloak/saml/processing/api/saml/v2/sig/SAML2Signature.java:
```diff
@@ -210,18 +208,20 @@ public Node getNextSiblingOfIssuer(Document doc) {
         }
     }
+    
+    public static void configureIdAttribute(Element assertion) {
```
This method works for any element, not just assertion, please rename parameter accordingly.

hmlnarik commented:
I see this PR having a good point in fixing the SAMLEndpoint condition and some refactorings. Let's stick to these. Please leave the resolver changes out and create separate JIRA for them to have them fixed properly in a separate PR.

hmlnarik reviewed (changes requested):
.

skjolber commented:
Ok, so lets go with the resolver first, before finalizing this PR. I've created https://issues.jboss.org/browse/KEYCLOAK-4979, but am not sure where to copy the picketlink schema. saml-core is my best suggestion.

skjolber commented on saml-core/src/main/java/org/keycloak/saml/processing/api/saml/v2/sig/SAML2Signature.java:
```diff
@@ -210,18 +208,20 @@ public Node getNextSiblingOfIssuer(Document doc) {
 
         for (int i = 0; i < nodes.getLength(); i++) {
-            Node n = nodes.item(i);
-            if (n instanceof Element) {
```
document.getElementsByTagNameNS(...) returns only Element, so the if statement is always true.

hmlnarik replied:
That's true according to the documentation, but in reality it returns an untyped NodeList where by mistake a non-Element node can occur. Better safe than sorry here, please keep the condition.

skjolber replied:
If the Document implementation is not according to spec, and the cast to Element fails, we then would be aware of and able to fix the bug in the implementation. If we do not trust the implementation to work, why bother calling the method - it might 'not work' in countless ways. The code should be safe, as there is no logical loophole (an exception will always occor). 

So unless there is a concrete example of an implemenation which triggers an exception in the code, on which the project is dependant and which cannot be fixed, keeping the old code does not make sense.

hmlnarik commented:
I appreciate your will to fix the resolver but that is of lower priority than this PR. Please fix this PR as requested.

skjolber commented:
This PR has been updated as requested.

hmlnarik commented on services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java:
```diff
@@ -371,13 +371,12 @@ protected Response handleLoginResponse(String samlResponse, SAMLDocumentHolder h
-                    }
+                if ((config.isWantAssertionsSigned() && !AssertionUtil.isSignedElement(assertionElement))
+                        || (config.isValidateSignature() && !AssertionUtil.isSignatureValid(assertionElement, getIDPKeyLocator()))) {
+                    logger.error("validation failed");
```
Would this condition work for SAML documents where only the whole document (not the assertions) is signed?
Please create a KcSamlSignedDocumentOnlyBrokerTest based on org.keycloak.testsuite.broker.KcSamlSignedBrokerTest, check https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/KcSamlSignedBrokerTest.java#L52.

skjolber replied:
Does signing the whole document cound as signing the assertion?

hmlnarik replied:
I don't understand your question. The changed code checks for validity of Assertion element regardless of the wantAssertionsSigned flag. The IdP can be configured to sign the whole document only, not individual assertions. Is this case still working? The PR has to contain an integration test for that.

skjolber replied:
Okey, I understand better now.

skjolber replied:
If validateSignature is true, should the signature be check regardless of whether the assertion was wanted or not?

hmlnarik commented on saml-core/src/main/java/org/keycloak/saml/processing/core/util/XMLSignatureUtil.java:
```diff
@@ -409,7 +409,7 @@ public static Document sign(SignatureUtilTransferObject dto, String canonicaliza
         return doc;
     }
-
+    
```
Please remove this change.

hmlnarik reviewed (changes requested):
Thank you. Please see comments in individual lines

hmlnarik commented:
@skjolber When do you expect to update this PR?

skjolber commented:
@hmlnarik Made some adjustments now, but frankly not sure the KcSamlSignedDocumentOnlyBrokerTest is ready.

hmlnarik commented:
> Made some adjustments now, but frankly not sure the KcSamlSignedDocumentOnlyBrokerTest is ready.
Why? It seems correct to me.

> Edit: So the validateSignature setting, does it imply that Assertion signatures should be checked, regardless of whether they were expected (wanted) or not?
Yes.

hmlnarik reviewed (approved):
@skjolber Thank you
@stianst This PR can be merged.

skjolber commented:
@hmlnarik Thanks for reviewing!
